cardano-testnet/src/Testnet/Property/Util.hs

```diff
@@ -60,18 +63,44 @@ integrationWorkspace workspaceName f = withFrozenCallStack $
 isLinux :: Bool
 isLinux = os == "linux"
 
-
 -- | Runs an action in background, and registers cleanup to `MonadResource m`
--- The argument forces IO monad to prevent leaking of `MonadResource` to the child thread
+-- Concurrency is tricky in the 'ResourceT' monad. See the "Concurrency" section of
+-- https://www.fpcomplete.com/blog/understanding-resourcet/.
 runInBackground :: MonadTest m
                 => MonadResource m
-                => MonadCatch m
-                => IO a
+                => IO ()
+                -> IO a
                 -> m ()
-runInBackground act = void . H.evalM $ allocate (H.async act) cleanUp
-  where
-    cleanUp :: H.Async a -> IO ()
-    cleanUp a = H.cancel a >> void (H.link a)
+runInBackground runOnException act  =
+  void . H.evalIO
+       $ runResourceT
+       -- We don't 'wait' because this "background process" may not terminate.
+       -- If we 'wait' and it doesn't terminate, 'ResourceT' will not kill it
+       -- and the test will hang indefinitely.
+       -- Not waiting isn't a problem because this "background process"
+       -- is meant to run indefinitely and will be cleaned up by
+       -- 'ResourceT' when the test ends or fails.
+       -- We use 'asyncWithUnmask' because our logging thread is terminated via an exception.
+       -- In order to avoid competing for a file handle we must catch the exception which signals
+       -- the logging file is no longer being written to and we can now run the desired additional IO action we
+       -- want (runOnException). Attempting to share the 'FileHandle' and use concurrency primitives was not fruitful
+       -- and the section "Other ways to abuse ResourceT" in https://www.fpcomplete.com/blog/understanding-resourcet/
+       -- confirms this is problematic in 'ResourceT'.
+       $ resourceForkWith (\_ -> do r <- H.asyncWithUnmask (\restore -> restore act `E.onException` runOnException)
+                                    linkOnly ignoreException r
+                          ) $ return ()
+ where
+  ignoreException  :: E.SomeException -> Bool
+  ignoreException e =
+    case E.fromException e of
+     Just (MuxError errType _) ->
```

Comment:
Can you show me what you want here? We need to ignore the `MuxBearerClosed` which occurs because the node is `foldEpochState` is connected to is shut down.

Reply:
I'd like `foldEpochState` to return an error in `ExceptT` or a synchronous exception - not both of them. We're catching `IOExceptions` there already for example.